src/tensors/dbcsr_tensor_api_c.F

```diff
@@ -0,0 +1,1231 @@
+      TYPE(dbcsr_t_type), POINTER                        :: tensor
+
+      ALLOCATE (iterator)
+      CALL c_f_pointer(c_tensor, tensor)
+
+      CALL dbcsr_t_iterator_start(iterator, tensor)
+
+      c_iterator = c_loc(iterator)
+
+   END SUBROUTINE
+
+
+   SUBROUTINE c_dbcsr_t_iterator_stop(c_iterator) BIND(C, name="c_dbcsr_t_iterator_stop")
+
+      TYPE(c_ptr), INTENT(INOUT)                         :: c_iterator
+      TYPE(dbcsr_t_iterator_type), POINTER               :: iterator
+
+      CALL c_f_pointer(c_iterator, iterator)
+
+      CALL dbcsr_t_iterator_stop(iterator)
+
+      DEALLOCATE (iterator)
+
+      c_iterator = c_null_ptr
+
+   END SUBROUTINE
+
+
+   SUBROUTINE c_dbcsr_t_iterator_next_block(c_iterator, iterator_size, c_ind_nd, c_blk, c_blk_p, c_blk_size, c_blk_offset) &
+      BIND(C, name="c_dbcsr_t_iterator_next_block")
+
+      TYPE(c_ptr), INTENT(IN), VALUE                     :: c_iterator
+      INTEGER(kind=c_int), INTENT(IN), VALUE             :: iterator_size
+      INTEGER(kind=c_int), DIMENSION(iterator_size), TARGET, &
+         INTENT(INOUT)                                   :: c_ind_nd
+      INTEGER(kind=c_int), INTENT(INOUT)                 :: c_blk
+      INTEGER(kind=c_int), INTENT(INOUT), OPTIONAL       :: c_blk_p
+      INTEGER(kind=c_int), DIMENSION(iterator_size), INTENT(INOUT), &
+         TARGET, OPTIONAL                                :: c_blk_size, c_blk_offset
+
+      TYPE(dbcsr_t_iterator_type), POINTER               :: iterator
+      INTEGER, DIMENSION(:), POINTER                     :: ind_nd => null()
+      INTEGER, ALLOCATABLE                               :: blk_p
+      INTEGER, DIMENSION(:), POINTER                     :: blk_size => null()
+      INTEGER, DIMENSION(:), POINTER                     :: blk_offset => null()
+
+      CALL c_f_pointer(c_iterator, iterator)
+      ind_nd => c_ind_nd
+
+      IF (PRESENT(c_blk_p)) THEN 
+        ALLOCATE(blk_p) 
+        blk_p = c_blk_p
+      ENDIF
+      IF (PRESENT(c_blk_size)) blk_size => c_blk_size
+      IF (PRESENT(c_blk_offset)) blk_offset => c_blk_offset
+
+      CALL dbcsr_t_iterator_next_block(iterator, ind_nd, c_blk, blk_p, blk_size, blk_offset)
+      
+      IF (ALLOCATED(blk_p)) DEALLOCATE(blk_p)
+
+      c_ind_nd = c_ind_nd - 1
+      
+      NULLIFY(ind_nd)
+
+   END SUBROUTINE
+
+
+   FUNCTION c_dbcsr_t_iterator_blocks_left(c_iterator) &
+      BIND(C, name="c_dbcsr_t_iterator_blocks_left")
+
+      TYPE(c_ptr), INTENT(IN), VALUE                     :: c_iterator
+      TYPE(dbcsr_t_iterator_type), POINTER               :: iterator
+      LOGICAL(kind=c_bool)                               :: c_dbcsr_t_iterator_blocks_left
+
+      CALL c_f_pointer(c_iterator, iterator); 
+      c_dbcsr_t_iterator_blocks_left = dbcsr_t_iterator_blocks_left(iterator)
+
+   END FUNCTION
+
+
+   SUBROUTINE c_dbcsr_t_split_blocks(c_tensor_in, tensor_dim, c_tensor_out, c_block_sizes, c_nodata) &
```

Comment:
This routine should not be in the API (my bad, I see it's in the Fortran API and used within CP2K), `dbcsr_t_copy` should do all the tricks. We can leave it for now but I may remove it later.

Reply:
Ok.